File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/H-Хорошие_дни/solution.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    scanf("%d", &n);
    vector<long long> prefixSum = {0};
    vector<int> nums;
    for (int i = 0; i < n; ++i) {
        int num;
        scanf("%d", &num);
        prefixSum.push_back(prefixSum.back() + num);
        nums.push_back(num);
    }
    stack<int> st;
    vector<pair<int, int>> segments;
    for (int idx = 0; idx < n; ++idx) {
        while (!st.empty() && nums[st.top()] >= nums[idx]) {
            segments[st.top()].second = idx;
            st.pop();
        }
        if (st.empty()) segments.push_back({-1, n});
        else segments.push_back({st.top(), n});
        st.push(idx);
    }
    long long answer = 0;
    for (int i = 0; i < n; ++i) {
        int left = segments[i].first + 1;
        int right = segments[i].second;
        long long part = (prefixSum[right] - prefixSum[left]) * nums[i];
        answer = max(answer, part);
    }
    printf("%lld\n", answer);
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/H-Хорошие_дни/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    sequence: iter[int] = map(int, sys.stdin.readline().rstrip().split())
    prefix_sum: list[int] = [0]
    nums: list[int] = []
    for _ in range(n):
        num: int = next(sequence)
        prefix_sum.append(prefix_sum[-1] + num)
        nums.append(num)
    
    stack: list[int] = []
    segments: list[list[int, int]] = []
    for idx in range(n):
        while stack and nums[stack[-1]] >= nums[idx]:
            segments[stack.pop()][1] = idx
        if not stack: segments.append([-1, n])
        else: segments.append([stack[-1], n])
        stack.append(idx)
    answer: int = 0
    for i in range(n):
        left: int = segments[i][0] + 1 
        right: int = segments[i][1]
        part: int = (prefix_sum[right] - prefix_sum[left]) * nums[i]
        answer = max(answer, part)
    sys.stdout.write(f'{answer}\n')


if __name__ == '__main__':
    solution()